src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,166 @@
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexRange(index);
+        Node<T> foundNodeByIndex = findNodeByIndex(index);
+        T oldElement = foundNodeByIndex.element;
+        foundNodeByIndex.element = value;
+        return oldElement;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexRange(index);
+        Node<T> removedNode = findNodeByIndex(index);
+        unlink(removedNode);
+        size--;
+        return removedNode.element;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> removedNodeByValue = findNodeByValue(object);
+        if (removedNodeByValue == null) {
+            return false;
+        } else if (size == 0) {
+            head = null;
+        } else if (size == 1) {
+            tail = null;
+        }
+        unlink(removedNodeByValue);
+        size--;
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexRange(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: "
+                    + index
+                    + " out of bounds "
+                    + "for size "
+                    + size);
+        }
+    }
+
+    private void checkIndexRangeInAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: "
+                    + index
+                    + " out of bounds "
+                    + "for size "
+                    + size);
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private Node<T> findNodeByValue(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if ((currentNode.element == object)
+                    || currentNode.element != null
+                    && currentNode.element.equals(object)) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+        }
+        return null;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev == null) {
+            head = head.next;
+        } else if (node.next == null) {
+            tail = tail.prev;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+    }
+
+    private class Node<E> {
```

Comment:
```suggestion
    private static class Node<E> {
```
prefer static for inner classes where possible